### PR TITLE
[SDK-152] Feature/add webview events

### PR DESCRIPTION
**Pull request by HarrisonHough**
- this PR adds support for all the new web browser events that come from RPM
- it also includes a big refactor and rework of the WebBrowser integration 
Specific changes include:
- javascript snippet moved to a separate .js file to make it easier to edit, this snippet is loaded at runtime as a string
- New delegates have been added to the ReadyPlayerMeWebBrowser.cpp class which can be subscribed to from blueprints as well as C++ 
- some data types added to hold the data from the web events

### CODE REVIEW

**Comment by gadamyan** [comment]
Now that we added more C++ files, maybe it makes sense to convert it to an optional plugin.

File: Source/UnrealExamples/Public/WebMessage.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#pragma once
+
+#include "CoreMinimal.h"
+
+/**
+ * 
+ */
+class UNREALEXAMPLES_API WebMessage
```
Comment (gadamyan) [inline Source/UnrealExamples/Public/WebMessage.h:7]: What is this class for ?

File: Config/DefaultEngine.ini [hunk 2]
```diff
@@ -47,3 +47,6 @@ ConnectionType=USBOnly
 bUseManualIPAddress=False
 ManualIPAddress=
 
+
+[CoreRedirects]
++FunctionRedirects=(OldName="/Script/UnrealExamples.WebLink.SetAvatarUrlCallback",NewName="/Script/UnrealExamples.WebLink.SetResponseCallback")
```
Comment (gadamyan) [inline Config/DefaultEngine.ini]: This is usually necessary if someone else modifies the UnrealExamples project, and then tries to fetch the latest changes from git. I'd remove this, in theory, no one should modify the project and then update it.

File: Content/ReadyPlayerMe/WebBrowser/Scripts/rpmFrameSetup.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+function setupEventSubscription() {
```
Comment (gadamyan) [inline Content/ReadyPlayerMe/WebBrowser/Scripts/rpmFrameSetup.js]: Could you please check the Unreal guidelines, if it's OK for unreal that the file name starts with the lower case.

Comment (HarrisonHough) [reply Content/ReadyPlayerMe/WebBrowser/Scripts/rpmFrameSetup.js]: I can check, the reason I left it lowercase is it's kinda standard for JS. 
I'll just double check using a capital doesn't cause issues (I think it won't) and I'll update

File: Source/UnrealExamples/Public/WebMessage.h (hunk 1 above)
Comment (HarrisonHough) [reply Source/UnrealExamples/Public/WebMessage.h:7]: right check again now, I moved the WebMessage from a struct to this as it had more functionality then a struct should have

File: Config/DefaultEngine.ini (hunk 2 above)
Comment (HarrisonHough) [reply Config/DefaultEngine.ini]: yep this was auto-added by rider I wasn't sure what it did :D will remove

File: Source/UnrealExamples/Public/WebViewEvents.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	UPROPERTY(BlueprintReadWrite, EditAnywhere, Category = "Ready Player Me|Web Browser")
+	FString UserId;
+
+	UPROPERTY(BlueprintReadWrite, EditAnywhere, Category = "Ready Player Me|Web Browser")
+	FString AssetId;
+};
+
+DECLARE_DYNAMIC_MULTICAST_DELEGATE_OneParam(FOnAvatarCreated, const FString&, AvatarUrl);
+
+DECLARE_DYNAMIC_MULTICAST_DELEGATE_OneParam(FOnUserSet, const FString&, userId);
```
Comment (gadamyan) [inline Source/UnrealExamples/Public/WebViewEvents.h]: Param name with upper case, for this and the next two params.

File: Source/UnrealExamples/UnrealExamples.Build.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@ public UnrealExamples(ReadOnlyTargetRules Target) : base(Target)
 	{
 		PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs;
 
-		PublicDependencyModuleNames.AddRange(new string[] { "Core", "CoreUObject", "Engine", "InputCore", "HeadMountedDisplay" });
+		PublicDependencyModuleNames.AddRange(new string[] { "Core", "CoreUObject", "Engine", "InputCore", "HeadMountedDisplay", "Json", "JsonUtilities" });
```
Comment (gadamyan) [inline Source/UnrealExamples/UnrealExamples.Build.cs]: these should be added to the PrivateDependencyModuleNames instead

File: Source/UnrealExamples/WebLink.h [hunk 6]
```diff
@@ -3,27 +3,22 @@
...
+	void SetWebBrowser(class UReadyPlayerMeWebBrowser& InWebBrowser)
+	{
+		WebBrowser = &InWebBrowser;
+	}
 
 private:
-
-	FReadyPlayerWebBrowserResponse WebBrowserResponse;
-	FString LastAvatarUrl;
+	UReadyPlayerMeWebBrowser* WebBrowser;
```
Comment (gadamyan) [inline Source/UnrealExamples/WebLink.h]: This should be declared a UPROPERTY() for the proper garbage collection.

File: Source/UnrealExamples/WebLink.h [hunk 7]
```diff
@@ -3,27 +3,22 @@
...
 
-	UFUNCTION(BlueprintCallable, meta = (DisplayName = "RPM Set Callback"), Category = "Ready Player Me")
-	void SetAvatarUrlCallback(const FReadyPlayerWebBrowserResponse& WebBrowserCallback);
+public:
+	UFUNCTION(BlueprintCallable, Category = "Ready Player Me|Web Browser")
+	void EventReceived(const FString JsonResponse);
 
-	UFUNCTION(BlueprintCallable, meta = (DisplayName = "RPM Get Last Avatar Url"), Category = "Ready Player Me")
-	FString GetLastAvatarUrl();
+	void SetWebBrowser(class UReadyPlayerMeWebBrowser& InWebBrowser)
```
Comment (gadamyan) [inline Source/UnrealExamples/WebLink.h]: pass a pointer instead `class UReadyPlayerMeWebBrowser* InWebBrowser`

File: Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp [hunk 8]
```diff
@@ -15,36 +20,84 @@ static const TCHAR* GenderFemaleParam = TEXT("gender=female");
...
 }
 
 void UReadyPlayerMeWebBrowser::BindBrowserToObject()
 {
-	if (!WebLinkObject) {
+	if (!WebLinkObject)
+	{
 		WebLinkObject = NewObject<UWebLink>(this, *LinkObjectName);
 	}
+	WebLinkObject->SetWebBrowser(*this);
```
Comment (gadamyan) [inline Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp]: Now that the UWebLink doesn't have much logic maybe it's better to bind the event to the WebBrowser instead?
```
WebBrowserWidget->BindUObject(LinkObjectName, this);
```
then move the `void EventReceived(const FString JsonResponse);` from the UWebLink to UReadyPlayerMeWebBrowser

File: Source/UnrealExamples/Private/WebViewEvents.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+#include "UnrealExamples/Public/WebViewEvents.h"
+#include "WebMessage.h"
+#include "Dom/JsonObject.h"
+
+const FString WebViewEvents::AVATAR_EXPORT = TEXT("v1.avatar.exported");
+const FString WebViewEvents::USER_SET = TEXT("v1.user.set");
+const FString WebViewEvents::USER_AUTHORIZED = TEXT("v1.user.authorized");
+const FString WebViewEvents::ASSET_UNLOCK = TEXT("v1.asset.unlock");
+
+const FString DataField = "data";
```
Comment (gadamyan) [inline Source/UnrealExamples/Private/WebViewEvents.cpp]: Constants need to be either static or under a namespace

File: Source/UnrealExamples/Private/WebMessage.cpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "WebMessage.h"
+#include "WebViewEvents.h"
+
+const char* KeyId = "id";
```
Comment (gadamyan) [inline Source/UnrealExamples/Private/WebMessage.cpp]: Constants need to be either static or under a namespace

File: Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp (hunk 8 above)
Comment (gadamyan) [reply Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp:64]: Wait I think you can not WebBrowserWidget doesn't inherit from the UObject.

File: Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp [hunk 11]
```diff
@@ -15,36 +20,84 @@ static const TCHAR* GenderFemaleParam = TEXT("gender=female");
...
-void UReadyPlayerMeWebBrowser::SetupBrowser(const FReadyPlayerWebBrowserResponse& Response)
+void UReadyPlayerMeWebBrowser::SetupBrowser()
 {
 	BindBrowserToObject();
-	WebLinkObject->SetAvatarUrlCallback(Response);
-	ExecuteJavascript(JSAddAvatarGeneratedListener);
+	const FString Path = FPaths::ProjectContentDir() / JavascriptPath;
+
+
+	FString rpmSetupJavascript;
```
Comment (gadamyan) [inline Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp]: Upper case

File: Source/UnrealExamples/Public/WebMessage.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#pragma once
+
+#include "CoreMinimal.h"
+
+class UNREALEXAMPLES_API WebMessage
+{
+public:
+	FString Type;
```
Comment (gadamyan) [inline Source/UnrealExamples/Public/WebMessage.h:10]: is this field used?

File: Source/UnrealExamples/Public/WebMessage.h [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+	FString Source;
```
Comment (gadamyan) [inline Source/UnrealExamples/Public/WebMessage.h]: is this field used?

File: Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp (hunk 8 above)
Comment (HarrisonHough) [reply Source/UnrealExamples/ReadyPlayerMeWebBrowser.cpp]: hmm good point, didn't think of that, will try it out and change if it works.

File: Source/UnrealExamples/Public/WebMessage.h (hunk 12 above)
Comment (HarrisonHough) [reply Source/UnrealExamples/Public/WebMessage.h:10]: it's not currently used but if a dev wants to check if it's a guest or logged in user they could use this

File: Source/UnrealExamples/Public/WebMessage.h (hunk 13 above)
Comment (HarrisonHough) [reply Source/UnrealExamples/Public/WebMessage.h]: Right source is only used javascript side, probably don't need to have it available to devs as it is currently always the same. (at least it should be) not sure if there are plans to change that

Comment (HarrisonHough) [reply Source/UnrealExamples/Public/WebMessage.h]: removed

**Review by gadamyan — APPROVED** [review]

File: Content/ReadyPlayerMe/WebBrowser/Scripts/rpmFrameSetup.js (hunk 3 above)
Comment (HarrisonHough) [reply Content/ReadyPlayerMe/WebBrowser/Scripts/rpmFrameSetup.js]: updated

**Review by gadamyan — APPROVED** [review]

**Pull request merged**